server/update-workflow-interface/update-workflow-interface.md

```diff
@@ -0,0 +1,217 @@
+# Proposal: Update Workflow Server Interface
+
+Currently (as of version 1.15), ad-hoc runtime interaction with Temporal
+workflows is achieved through queries and signals. The former technique can
+interrogate the state of a workflow and returns a result but must not _change_
+the state of the workflow, while the latter can change workflow state but does
+not natively support return values [^1]. 
+
+The missing piece in runtime workflow interaction is an affordance that
+provides both (1) the ability to change workflow state, and (2) a mechanism to
+observe the result of the same without polling or setting up callback
+structures. This feature will be workflow updates.
+
+This proposal describes the server API (i.e. gRPC) for such a feature. We leave
+implementation details for a deeper design review and instead relate the API,
+its expected usage and the guarantees provided.
+
+## Concepts
+
+### Update
+
+An Update is the execution of arbitrary workflow logic (subject to the standard
+workflow code constraints on determinism) triggered by an UpdateRequest and
+eventually resulting in a single UpdateResponse. The processing of an Update is
+performed by workflow code and as such the timeliness of the UpdateResponse is
+dependent on worker resources being available to process the Update. A given
+Update applies to exactly one workflow execution. The lifetime of an updated
+begins at the first UpdateRequest triggering said Update and ends when the
+UpdateResponse has been made durable. Consequently, the lifetime of an Update
+may exceed the lifetime of the RPC request/response that initiates it.
+
+### Update ID
+
+The unique identifier of an Update. Generated by the Temporal server.
+
+### UpdateRequest
+
+An UpdateRequest is a remote call originating either from outside the Temporal
+system entirely or from another Temporal workflow.  UpdateRequests are
+deduplicated such that issuance of the same UpdateRequest multiple times is
+effectively the same as issuing a single request. The UpdateRequest includes an
+indication of the workflow execution at which the Update is targeted.
+UpdateRequests are issued over gRPC or through SDK-specific APIs wrapping the
+same. Both issuing and receiving UpdateRequests becomes a durable part of a
+given workflow's history.
+
+### UpdateResponse
+
+The output value of an Update. The UpdateResponse may lack a data playload, in
+which case the UpdateResponse only indicates the success or failure of the
+Update operation. The UpdateResponse is accessed via gRPC or SDK-specific APIs
+wrapping the same. UpdateResponses become durable parts of a given workflow's
+history. UpdateResponses can be accessed after the lifecycle of the Update has
+completed, subject to history storage & archival configuration.
+
+## Use Cases
```

Comment:
I can think of three more use cases:

* Fire-and-forget
* Duplicate/reentrant updates (maybe with different payload?)
* UpdateWithStart - SignalWithStart, but with an update (we don't have to support this now, just a thought)
* As a signal replacement - From a caller POV, what is a signal but an update that you don't care the response of? Why would I ever use signals again?
* UpdateExternalWorkflow - From _inside_ a workflow updating another (not part of the client gRPC of course)

None of this needs to be solved now, just some idle thoughts. I haven't really been involved in any of the discussions around update so I am surely missing lots of context.

Reply:
:+1: Bonus if we can change the requirement that workflow code has to run before an update ID is even issued. Mentioned elsewhere, but like signals, I'd like not not care about the worker to get this into history for processing.

Reply:
Signal replacement makes a lot of sense to me. Reducing the number of primitives we have is always a nice thing.

Reply:
Updates could replace signals in some use cases.
There are tradeoffs to each approach.
Updates can be rejected by the workflow in which case they will not be persisted in history, this is useful when you need to validate the update before applying it and makes update a mechanism that could be exposed directly to clients.
Signals OTOH are better for async processing, they don't require a worker to be up to be processed, they're much cheaper and involve less bookkeeping.

UpdateExternalWorkflow is hinted above in the `UpdateRequest` section.
UpdateWithStart was discussed too, we will definitely have that.

Reply:
>  UpdateWithStart was discussed too, we will definitely have that.

When we do get there, I'd like to propose we instead just add fields to `StartWorkflowExecutionRequest` instead of new RPCs (same for signal really, but of course we already have something there). Can I send a signal _and_ 3 updates to start the workflow with? :-)

Reply:
We decided a long time ago not to use "switch structures" to decide which operation is implemented by the given gRPC method. It was done when Thrift was used instead of gRPC which didn't have explicit support for "oneof" types. For consistency I would keep that pattern unless we want to refactor all the other APIs. For example Complete/Fail activity, etc.